pkg/controllers/jobset_controller.go

```diff
@@ -577,9 +578,10 @@ func constructJob(js *jobset.JobSet, rjob *jobset.ReplicatedJob, jobIdx int) (*b
 	labelAndAnnotateObject(&job.Spec.Template, js, rjob, jobIdx)
 
 	// If enableDNSHostnames is set, update job spec to set subdomain as
-	// job name (a headless service with same name as job will be created later).
-	if dnsHostnamesEnabled(rjob) {
-		job.Spec.Template.Spec.Subdomain = GenSubdomain(js, rjob)
+	// the chosen subdomain name or the default name (i.e. <jobsetName>).
+	// A headless service with this name will be created if does not exist.
+	if dnsHostnamesEnabled(js) {
+		job.Spec.Template.Spec.Subdomain = GenSubdomain(js)
```

Comment:
webhook defaulting should handle this

Reply:
In testing I found it didn't (the string was empty for subdomain) so I kept it for now. I also like the design that the GenSubdomain function is the single source of truth for creating the jobs and service.

Reply:
The webhook doesn't run during controller integration tests, it only runs for the webhook integration tests and e2e tests, so the existing controller integration test cases will fail if you remove this since the webhook defaulting won't be occurring. 

We might be able to run the webhook during controller integration tests by adding the `WebhookInstallOptions` to the testEnv setup like we do [here](https://github.com/kubernetes-sigs/jobset/blob/main/test/integration/webhook/suite_test.go#L67-L68) in the webhook integration tests, but I haven't tried it.

Reply:
Right, we shouldn't have defaulting in two places. For the unit tests here, just explicitly set the domain.

I would also log an error if dnsHostnamesEnabled is true and subdomain is empty.